### ISSUE TITLE
unbindAll doesn't work as expected

### ISSUE THREAD

**Issue by zkyronex** [pusher/pusher-websocket-swift#116]
Hi,

I have a reference to channel as returned by `pusher.subscribe(channelName)` which binds to a few events.
When I no longer want the event callbacks to trigger, I call `channel.unbindAll()`. I also call `pusher.unsubscribe(channelName)` and `pusher.disconnect()`

However, when I resubscribe to the channel and events, the callback triggers twice.
Perhaps the way I am handling the pusher object's lifecycle is incorrect.

Best regards.

**Comment by hamchapman** [comment]
Do you have a sample project or if that's not possible, just a copy and pasted version of the Pusher-related code that's leading to this behaviour? Thanks!

**Comment by zkyronex** [comment]
Below is a snippet of code I use to enable pusher so you get the idea. I basically put all my logic into the 2 functions: start and stop. I bind using channel as I am not interested in global events.

```
let pusher = Pusher(key: PusherKeys.apiKey)
private var channel: PusherChannel?

// called during viewWillAppear()
func startListening() {
        let channel = pusher.subscribe(PusherKeys.channel)
        
        channel.bind(eventName: PusherKeys.addItem) { [weak self] data in
            guard
                let data = data,
                let post = self?.parseAddItem(data: data)
                else { return }
            
            // handle add
        }
        
        channel.bind(eventName: PusherKeys.removePost) { [weak self] data in
            guard
                let data = data,
                let postID = self?.parseRemovePost(data: data)
                else { return }
            
            // handle remove
        }
        
        self.channel = channel
        pusher.connect()
    }
    
// called during viewDidDissapear()
func stopListening() {
        channel?.unbindAll()
        
        let channelName = PusherKeys.channel
        pusher.unsubscribe(channelName)
        pusher.disconnect()
    }

```

**Comment by thefredelement** [comment]
I experience exactly the same, only after managing state (pusher.connect/disconnect) and only seemingly once per a channel.

**Comment by hamchapman** [comment]
Hi guys, I've tried to recreate this locally (literally copy and pasted the code @zkyronex added above) and am unable to get duplicate closures being called on message receipt.

Are you able to provide a full sample project that exhibits this behaviour?

**Comment by thefredelement** [comment]
I'm under a deadline for Friday but will if I have time. I *think* the issue is that channels/event call backs don't use thread safe getters/setters.

As a workaround I cast the payload data into a dictionary and retain that dictionaries hashValue property, if the next message matches it, the call back block returns. (Doesn't work really well on ARM32)

**Comment by zkyronex** [comment]
My scenario seems to be less complicated than @thefredelement. When I revisited this issue, I found that my preparation and binding setup was incorrect. I cannot replicate this issue anymore. Apologies for not updating everyone on this sooner.

@hamchapman Could you advise (maybe put some documentation in) on handling the lifecycle and setup of the Pusher instance (connections and channels) as well as how to clean up and tear down when no longer needed.

Also, something noteworthy is the internal `connectionState` handling which goes on within PusherSwift. 
As the state updates rely on the `Starscream` WebSocketDelegates, calling `connect()` and `disconnect()` may not trigger the expected behaviour all the time.
The `connect()` function could guard against an extra state of `.connecting` to avoid duplicate connection calls.
The `disconnect()` function could execute on `.connected` as well as `.connecting` to stop the disconnect() function from failing without getting handled (maybe during viewcontroller.viewDidDissapear)

The above are merely my opinions based of what I managed to observe by stepping through and testing the library. Happy to hear back and if there is something I could do to improve the library, I'd be happy to make a PR. 
Cheers!
